File: 209. Minimum Size Subarray Sum/optimised.js
/**
 * @param { number } target
 * @param { number[] } nums
 * @return { number }
 */

var minSubArrayLen = function(target, nums) {
	const n = nums.length;
	let left = 0;
	let total = 0;
	let ans = Infinity;

	for (let right = 0; right <= n; right++) {
		total += nums[right];

		while (total >= target) {
			ans = Math.min(ans, right - left + 1);
			total -= nums[left];
			left += 1;
		}
	}

	return ans === Infinity ? 0 : ans;
};

console.log(minSubArrayLen(4, [2,3,1,2,4,3]));

File: 209. Minimum Size Subarray Sum/optimised.py
#!/usr/bin/env python

from typing import List

def minSubArrayLen(target: int, nums: List[int]) -> int:
    n = len(nums)
    left = 0
    total = 0
    ans = float('inf')

    for right in range(n):
        total += nums[right]

        while total >= target:
            ans = min(ans, right - left + 1)
            total -= nums[left]
            left += 1

    return 0 if ans == float('inf') else ans

print(minSubArrayLen(4, [1,4,4]))

print(minSubArrayLen(11, [1,1,1,1,1,1,1,1]))
